### ISSUE TITLE
"copy --repair" duplicates sectors if no offset data are present in one disk image

### ISSUE THREAD

**Issue by shattered** [simonowen/samdisk#11]
I have two images of a disk - a td0 and a scp; td0 has one bad sector which is OK in scp image.  Disk itself is not available.   But td0 stores no track length or offset data, so copy --repair doesn't merge identical sectors:

> 250Kbps MFM, 20 sectors,  512 bytes/sector:
>   0.0  1[r] 2[r] 3[r] 4[r] 5[r] 6[r] 7[r] 8[r] 9[r] 10[r] 1[r] 2[r] 3[r] 4[r] 5[r] 6[r] 7[r] 8[r] 9[r] 10[r]

**Comment by simonowen** [comment]
Add is usually called when creating new disks, where the source type remains the same.  I hadn't thought about repair mixing offset and non-offset disks -- good spot!

I have some other improvements that could probably be included at the same time.

**Comment by simonowen** [comment]
This should be fixed by 04af97a2ab47e98cf98198c995fa41c504fe118c

Thanks for the bug report!

**Comment by shattered** [comment]
Thanks, new messages from --repair are useful -- I've written a quick wrapper to run it with a range of values for pll-phase/pll-adjust/scale options to try and recover as much as possible.

But repairing with a .scp source still duplicates sectors...

**Comment by simonowen** [comment]
v3 used to show the same messages, so I think it’s good to have them back.

I’m surprised that you’re still seeing duplicate sectors as the new code selectively uses insert(), which doesn’t rely on positions.

Could the encoding be different? TD0 will only store MFM or FM, and if the SCP has Agat they’ll be considered distinct.

Does using —encoding=X to force a type help? The source and target are normalised, which would mean they’re the same before the repair.

**Comment by shattered** [comment]
This SCP has MFM (same as matching TD0).   I can upload both of them.

**Comment by simonowen** [comment]
Sample images would be useful, thanks.  One extra question: .scp and .td0 are both read-only, so could you give the commands you're using too?

I've successfully repaired a .dsk file with a .scp image, without seeing duplicate sectors, so I'm still missing where the problem is.

**Comment by shattered** [comment]
[issue11.zip](https://github.com/simonowen/samdisk/files/1817470/issue11.zip)

a1.dsk is OK, a2.dsk has duplicated sectors.

```
samdisk copy ARCSAV.TD0 a1.dsk && cp a1.dsk a2.dsk
samdisk copy disk4-1.scp b.dsk

samdisk copy b.dsk a1.dsk --repair --no-gaps
Fixed: repaired cyl 67 head 1 sector 6

samdisk copy disk4-1.scp a2.dsk --repair --no-gaps
Fixed: added missing cyl 0 head 0 sector 1
<...>
```

**Comment by simonowen** [comment]
Thanks for the sample files.  The issue is fixed in 115b02e3cd7814f553bef4253fe7f8bad7df008d.

The TD0 was seen as 250Kbps, and the SCP as 300Kbps.  The difference in data rate meant the sectors didn't match, so duplicates were added as though they were missing from the target.  Mixed data rates within a track is invalid and caused other problems, so it now also fails attempts to do that.

The solution was to be more flexible with 250/300Kbps sectors, so they are treated as equivalent.  I've also improved the merge status so adding more gap data doesn't report that a sector was improved. The result with your files is now just the single bad sector being repaired.
